fairlearn/metrics/_disaggregated_result.py

```diff
@@ -351,49 +324,62 @@ def create(
-            overall = apply_to_dataframe(data, metric_functions=annotated_functions)
-        else:
-            temp = data.groupby(by=control_feature_names).apply(
-                apply_to_dataframe,
-                metric_functions=annotated_functions,
-                # See note in apply_to_dataframe about include_groups
-                include_groups=False,
-            )
-            # If there are multiple control features, might have missing combinations
-            if len(control_feature_names) > 1:
-                cf_classes = extract_unique_classes(data, control_feature_names)
-                all_indices = pd.MultiIndex.from_product(
-                    cf_classes.values(), names=cf_classes.keys()
-                )
+        overall = DisaggregatedResult._apply_functions(
+            data=data,
+            annotated_functions=annotated_functions,
+            grouping_names=control_feature_names,
+        )
 
-                overall = temp.reindex(index=all_indices)
-            else:
-                overall = temp
+        by_group = DisaggregatedResult._apply_functions(
+            data=data,
+            annotated_functions=annotated_functions,
+            grouping_names=(control_feature_names or []) + sensitive_feature_names,
+        )
+
+        return DisaggregatedResult(overall, by_group)
 
-        # Calculate the 'by_group' values
-        all_grouping_names = [x for x in sensitive_feature_names]
-        if control_feature_names is not None:
-            # Note that we prepend the control feature names
-            all_grouping_names = control_feature_names + all_grouping_names
+    @staticmethod
+    def _apply_functions(
+        *,
+        data: pd.DataFrame,
+        annotated_functions: Dict[str, AnnotatedMetricFunction],
+        grouping_names: Optional[List[str]],
+    ) -> Union[pd.Series, pd.DataFrame]:
+        """
+        Apply annotated metric functions to a DataFrame, optionally grouping by specified columns.
+
+        Parameters
+        ----------
+        data : pd.DataFrame
+            The input data on which the metric functions will be applied.
+        annotated_functions : Dict[str, AnnotatedMetricFunction]
+            A dictionary where keys are metric names and values are the corresponding annotated metric
+            functions.
+        grouping_names : Optional[List[str]]
+            A list of column names to group by before applying the metric functions. If None, the
+            functions are applied to the entire DataFrame.
 
-        temp = data.groupby(all_grouping_names).apply(
+        Returns
+        -------
+        Series or DataFrame
+            A Series or DataFrame with the results of the metric functions applied. If grouping_names is provided,
+            the results are grouped accordingly.
+        """
+        if grouping_names is None:
+            return apply_to_dataframe(data, metric_functions=annotated_functions)
+
+        assert len(grouping_names) > 0
```

Comment:
we should avoid `assert` in production code. If we need to do input validation, it should raise meaningful errors with nice error messages on how to fix the issue.

Reply:
Actually an empty `grouping_names` can be a valid case (equivalent to `None`). I fixed the function and added a corresponding test case